### PR TITLE
Fixing get_site_url bugs with solid tests

### PR DESCRIPTION
**Pull request by oudeismetis**
**Description:**
Bringing in some lessons learned I have from prior work.
Adding good test examples to help others with future testing

TODO:
- [x] Implement feature and/or fix bug
- [x] Add and/or fix tests
- [x] Review App looks good
- [x] Code Review

### CODE REVIEW

File: {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/sites.py [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
+    if not (settings.HEROKU_APP_NAME or settings.CURRENT_DOMAIN):
+        raise Exception("Environment variable for either HEROKU_APP_NAME or CURRENT_DOMAIN must exist")
+    domain = f"{settings.HEROKU_APP_NAME}.herokuapp.com" if settings.HEROKU_APP_NAME else settings.CURRENT_DOMAIN
+    domain = domain.strip("/").split("/")[-1]
+    scheme = "http://" if settings.IN_DEV else "https://"
+    site_url = urlparse(f"{scheme}{domain}")
+    if settings.CURRENT_PORT:
+        netloc = site_url.netloc + f":{settings.CURRENT_PORT}"
+        site_url = site_url._replace(netloc=netloc)
+    return site_url.geturl()
```
Comment (oudeismetis) [inline {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/sites.py:16]: The will now generate the correct URL for Heroku Review Apps
In addition, several edge case bugs are now handled. Don't have to worry about CURRENT_DOMAIN needing https are the beginning or not.

**Review by whusterj — APPROVED** [review]
LGTM 🎊  Tests look good. I ran them locally and they are passing - very good example for others, too.

QQ: Are these unit tests run by Github Actions?

I poked around the 'Checks' tab and I see where cookiecutter is tested, but I don't see where these tests are being run (if at all) or output from them.

File: {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/tests.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    [
+        ({"CURRENT_DOMAIN": None, "CURRENT_PORT": 8080, "HEROKU_APP_NAME": None, "IN_DEV": True}, "http://localhost:8080"),
+        ({"CURRENT_DOMAIN": "", "CURRENT_PORT": None, "HEROKU_APP_NAME": "", "IN_DEV": False}, "https://example.herokuapp.com"),
+    ],
+)
+def test_get_site_url_negative(settings, custom_settings, expected_output):
+    for key in custom_settings:
+        settings.__setattr__(key, custom_settings[key])
+    with pytest.raises(Exception):
+        get_site_url()
```
[the remark this answers is not in the document]
Comment (oudeismetis) [reply {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/tests.py:32]: great question! This is one of the oddities of pytest. In this case specifically `pytest-django` is injecting this fixture and making it available for all tests. So for any test to have access to the loaded settings or to manipulate them, you simple add it to the test like I did here.
And `parameterize` has the same oddity. You just have to KNOW that those key names will be the last X number of arguments for the function and that any other fixtures you are using have to come before.
I've been in situations where I'm doing what I've done here...but ALSO using several other fixtures, and it can get kludgy and confusing very quickly.

**Comment by oudeismetis** [comment]
Yes, the unit tests are run by Github Actions
image

You'll notice there are two broken out steps `Pytest_CookieCutter` and `Pytest_Generated_Project`. So separately we run any unit tests that are in the bootstrapper project (there are a couple) and then also generate a dummy `my_project` folder and run the tests that a developer who has bootstrapped an app would have.

We do the same for ESLint, Prettier, and Cypress. Only those don't need to run against the bootstrapper project and only run against the generated project.

File: {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/tests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (whusterj) [reply {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/tests.py:32]: I see - and is the ability to use `settings` as a param "built in" to pytest, or is it configured in the code somewhere?

[the remark this answers is not in the document]
Comment (oudeismetis) [reply {{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/tests.py:32]: built-in to `pytest-django` I believe. So it's a bit of magic that you wouldn't know about unless you read the docs or inspect all the fixtures that are initialized (which you can do, I forget how though)
